src/multi/content_identifier_codec.cpp

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <libp2p/multi/content_identifier_codec.hpp>
+#include <libp2p/multi/multicodec_type.hpp>
+#include <libp2p/multi/uvarint.hpp>
+
+OUTCOME_CPP_DEFINE_CATEGORY(libp2p::multi, ContentIdentifierCodec::EncodeError,
+                            e) {
+  using E = libp2p::multi::ContentIdentifierCodec::EncodeError;
+  switch (e) {
+    case E::INVALID_CONTENT_TYPE:
+      return "Content type does not conform the version";
+    case E::INVALID_HASH_LENGTH:
+      return "Hash length is invalid; Must be 32 bytes for sha256 in version 0";
+    case E::INVALID_HASH_TYPE:
+      return "Hash type is invalid; Must be sha256 in version 0";
+  }
+  return "Unknown error";
+}
+
+OUTCOME_CPP_DEFINE_CATEGORY(libp2p::multi, ContentIdentifierCodec::DecodeError,
+                            e) {
+  using E = libp2p::multi::ContentIdentifierCodec::DecodeError;
+  switch (e) {
+    case E::EMPTY_MULTICODEC:
+      return "Multicodec prefix is absent";
+    case E::EMPTY_VERSION:
+      return "Version is absent";
+    case E::MALFORMED_VERSION:
+      return "Version is malformed; Must be a non-negative integer";
+    case E::RESERVED_VERSION:
+      return "Version is greater than the latest version";
+  }
+  return "Unknown error";
+}
+
+namespace libp2p::multi {
+
+  outcome::result<std::vector<uint8_t>> ContentIdentifierCodec::encode(
+      const ContentIdentifier &cid) {
+    std::vector<uint8_t> bytes;
+    if (cid.version == ContentIdentifier::Version::V1) {
+      UVarint version(static_cast<uint64_t>(cid.version));
+      auto version_span = version.toBytes();
+      bytes.insert(bytes.end(), version_span.begin(), version_span.end());
```

Comment:
pls consider making a method like
```std::vector<uint8_t> &append(std::vector<uint8_t> &vec, gsl::span<uint8_t> span)```
and simplifying this code.

Reply:
Probably such method was already implemented somewhere on libp2p, pls look for possible implementation before creating your own.